lib/Transform/Clang/LoopWeightsEstimator.cpp

```diff
@@ -0,0 +1,219 @@
+  coverage::LineColPair MinEnclosingStartLoc{ 0, 0 };
+  coverage::LineColPair MinEnclosingEndLoc{ -1, -1 };
+  int64_t InstructionExecutionCount = 0;
+  for (const auto &CR : CRs) {
+    coverage::LineColPair CRStartLoc = CR.startLoc();
+    coverage::LineColPair CREndLoc = CR.endLoc();
+    if (CRStartLoc <= ILineCol && ILineCol <= CREndLoc
+        && MinEnclosingStartLoc < CRStartLoc && MinEnclosingEndLoc > CREndLoc) {
+      MinEnclosingStartLoc = CRStartLoc;
+      MinEnclosingEndLoc = CREndLoc;
+      InstructionExecutionCount = CR.ExecutionCount;
+    }
+  }
+  return InstructionExecutionCount;
+}
+
+uint64_t getInstructionEffectiveWeight(Instruction *I,
+    const std::vector<llvm::coverage::CountedRegion> &CRs,
+    TargetTransformInfo &TTI,
+    std::map<llvm::Function *, uint64_t> &FunctionEffectiveWeights) {
+  uint64_t InstructionExecutionCount = getExecutionCount(I, CRs);
+  uint64_t InstructionEffectiveWeight =
+      TTI.getInstructionCost(I, TargetTransformInfo::TCK_SizeAndLatency)
+      * InstructionExecutionCount;
+  Function *Called = nullptr;
+  if ((isa<CallInst>(I) || isa<InvokeInst>(I))
+      && (Called = cast<CallBase>(I)->getCalledFunction())
+      && !(Called->isIntrinsic())) {
+    uint64_t CalledEffectiveWeight = FunctionEffectiveWeights[Called];
+    uint64_t CalledExecutionCount = Called->getEntryCount().getCount();
+    InstructionEffectiveWeight +=
+        CalledEffectiveWeight * InstructionExecutionCount / CalledExecutionCount;
+  }
+  return InstructionEffectiveWeight;
+}
+}
+
+ClangLoopWeightsEstimator::ClangLoopWeightsEstimator() : ModulePass(ID) {
+  initializeClangLoopWeightsEstimatorPass(*PassRegistry::getPassRegistry());
+}
+
+bool ClangLoopWeightsEstimator::runOnModule(Module &M) {
+  auto &CRs = getAnalysis<ClangCountedRegionsExtractor>().getCountedRegions();
+  auto &CG = getAnalysis<CallGraphWrapperPass>().getCallGraph();
+  std::map<Function *, uint64_t> FunctionEffectiveWeights;
+  for (scc_iterator<CallGraph *> SCCIt = scc_begin(&CG); !SCCIt.isAtEnd(); ++SCCIt) {
+    for (auto *CGN : *SCCIt) {
+      Function *F = nullptr;
+      if ((F = CGN->getFunction()) && !(F->isIntrinsic())) {
+        auto &TTI = getAnalysis<TargetTransformInfoWrapperPass>().getTTI(*F);
+        std::map<Instruction *, uint64_t> InstructionEffectiveWeights;
+        uint64_t FunctionEffectiveWeight = 0;
+        for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+          uint64_t InstructionEffectiveWeight = getInstructionEffectiveWeight(&*I, CRs, TTI,
+              FunctionEffectiveWeights);
+          InstructionEffectiveWeights[&*I] = InstructionEffectiveWeight;
+          FunctionEffectiveWeight += InstructionEffectiveWeight;
+        }
+        FunctionEffectiveWeights[F] = FunctionEffectiveWeight;
+        bool *Changed = new bool();
+        *Changed = false;
+        auto &LI = getAnalysis<LoopInfoWrapperPass>(*F, Changed).getLoopInfo();
```

Comment:
```suggestion
        auto &LI = getAnalysis<LoopInfoWrapperPass>(*F).getLoopInfo();
```

Должна быть проверка, что у фнукции есть тело (`!F->empty()`), для функций без тела дерево циклов недоступно и попытка его получить может привести к падению программы.